### PR TITLE
DSDHacked [SOUNDS] support

### PR DESCRIPTION
**Pull request by electricbrass**
Implement PatchSounds for DSDHacked.

### CODE REVIEW

File: common/d_dehacked.cpp [hunk 1]
```diff
@@ -443,12 +444,31 @@ static const struct
...
-	{"[SPRITES]", PatchSprites},
     // Eternity engine added a few more features to BEX
     {"[MUSIC]", PatchMusic},
+	// DSDHacked BEX additions
+	{"[SPRITES]", PatchSprites},
+	{"[SOUNDS]", PatchSounds},
     {NULL, NULL},
 };
 
+DoomObjectContainer<const char*> SoundMap(ARRAY_LENGTH(doom_SoundMap));
```
Comment (cmbernard333) [inline common/d_dehacked.cpp:455]: I recommend putting this and its associated functionality into its own file `sounds.h/sounds.cpp` or something like that. That way anyone who wants this functionality can just include that header. Similar to `mobjinfo.h` and `sprites.h` etc.

File: common/d_dehacked.cpp [hunk 2]
```diff
@@ -443,12 +444,31 @@ static const struct
...
+
+void D_Initialize_SoundMap(const char** source, int count)
+{
+	SoundMap.clear();
+    if (source)
+    {
+		for (int i = 0; i < count; i++)
+		{
+			SoundMap.insert(source[i] ? strdup(source[i]) : NULL, i);
```
Comment (cmbernard333) [inline common/d_dehacked.cpp:464]: According to the [id24spec](https://drive.google.com/file/d/1zWebcl0gsjuH8VkWxNruS6gLzJGi9tIb/view?usp=drive_link), sfx has a `soundnum` that is int32_t. We should add those for all sounds if possible so we can reference them directly by index. This is similar to how mobjinfo and state_t have their own indices.

Comment (cmbernard333) [reply common/d_dehacked.cpp:464]: Though looking at PatchSounds DoomObjectContainer does have the ability to store and find the given index that a sound is stored with. Does this map resolve to an sfx struct somewhere?

File: common/d_dehacked.cpp [hunk 3]
```diff
@@ -1814,6 +1844,56 @@ static int PatchSprites(int dummy)
 	return result;
 }
 
+void IdxToSoundName(char** sound)
```
Comment (cmbernard333) [inline common/d_dehacked.cpp]: `const char*`?

File: common/d_dehacked.h [hunk 4]
```diff
@@ -223,7 +223,10 @@ static const char* SoundMap[] = { NULL,
                             // ZDOOM-Specific sounds
                             "misc/teamchat"};
 
+extern DoomObjectContainer<const char*> SoundMap;
```
Comment (cmbernard333) [inline common/d_dehacked.h:226]: Add to a header then this file can reference it (similar to mobjinfo).

File: common/d_dehacked.cpp (hunk 2 above)
Comment (electricbrass) [reply common/d_dehacked.cpp:464]: yes, the sound structs are stored in S_sfx, a vector that is mostly only modified and accessed by functions from `s_sound.cpp`

File: common/d_dehacked.cpp (hunk 3 above)
Comment (electricbrass) [reply common/d_dehacked.cpp]: Just changed this. Not sure why I did that before.

File: common/d_dehacked.cpp (hunk 1 above)
Comment (electricbrass) [reply common/d_dehacked.cpp:455]: I almost did this, but it seems like the original SoundMap was placed in `d_dehacked.h` for reason. It is only used with dehacked patches (vanilla and odamex mobjinfo have all sounds defined as the correct strings already), and would be included in exactly the same files that `d_dehacked.cpp` is already included in. To me this feels like it just complicates organization.

Comment (cmbernard333) [reply common/d_dehacked.cpp:455]: Seems logical to me if the SoundMap is only ever used by dehacked patches.

File: common/d_dehacked.cpp (hunk 2 above)
Comment (cmbernard333) [reply common/d_dehacked.cpp:464]: Making a note here (not necessary to implement right now): the struct channel_t in `s_sound.cpp` is similar to the id24spec struct `sfxinfo_s`. That's where we can do id24 work later on.

**Comment by electricbrass** [comment]
Don't want to merge this yet, it's causing some issues with multiplayer in eviternity 2.

**Comment by electricbrass** [comment]
> Don't want to merge this yet, it's causing some issues with multiplayer in eviternity 2.

Fixed, soundmap wasn't being initialized on the server.

**Pull request merged**